pkg/policy/api/rule_validation.go

```diff
@@ -37,19 +37,32 @@ func (r *Rule) Sanitize() error {
 		// Default posture is per-direction (ingress or egress),
 		// if there is a peer selector for that direction, the
 		// default is deny, else allow.
+		var hasEgress, hasIngress bool
 		if r.EnableDefaultDeny.Egress == nil {
-			x := len(r.Egress) > 0 || len(r.EgressDeny) > 0
-			r.EnableDefaultDeny.Egress = &x
+			hasEgress = len(r.Egress) > 0 || len(r.EgressDeny) > 0
+		} else {
+			hasEgress = *r.EnableDefaultDeny.Egress
 		}
+
 		if r.EnableDefaultDeny.Ingress == nil {
-			x := len(r.Ingress) > 0 || len(r.IngressDeny) > 0
-			r.EnableDefaultDeny.Ingress = &x
+			hasIngress = len(r.Ingress) > 0 || len(r.IngressDeny) > 0
+			if r.EnableDefaultDeny.Egress == nil && !hasEgress && !hasIngress {
+				return fmt.Errorf("rule must have one of Ingress/IngressDeny/Egress/EgressDeny/EnableDefaultDeny.Ingress/EnableDefaultDeny.Egress")
```

Comment:
I'm inclined to agree with Casey on this: If we declare a `oneOf` requirement in the CRD spec for the ingress/ingressDeny/egress/egressDeny then I don't think we need to change this runtime validation at all, because the misconfiguration would be caught in kube-apiserver.

This does subtly deviate from what you're proposing here, in that it makes the rule simpler: There must always be at least one rule in one of these statements. So rather than "supporting default deny policy with no rules", the proposal would "require at least one rule in every network policy".

Reply:
Hmm, with the [latest discussion point on the CFP issue](https://github.com/cilium/cilium/issues/35558#issuecomment-2503006466), @renyunkang points out a more complicated case that would require something like this runtime check. Not sure if there is a way to solve that problem purely with `kubebuilder` annotations.